aten/src/ATen/native/DilatedMaxPool2d.cpp

```diff
@@ -0,0 +1,499 @@
+    for (auto k = start; k < end; k++)
+    {
+      /* loop over output */
+      int64_t i, j;
+      scalar_t *ip = input_p   + k*iwidth*iheight;
+      for(i = 0; i < oheight; i++)
+      {
+        for(j = 0; j < owidth; j++)
+        {
+          int64_t hstart = i * dH - padH;
+          int64_t wstart = j * dW - padW;
+          int64_t hend = std::min(hstart + (kH - 1) * dilationH + 1, iheight);
+          int64_t wend = std::min(wstart + (kW - 1) * dilationW + 1, iwidth);
+          while(hstart < 0)
+            hstart += dilationH;
+          while(wstart < 0)
+            wstart += dilationW;
+
+          /* local pointers */
+          scalar_t *op = output_p  + k*owidth*oheight + i*owidth + j;
+          int64_t *indp = ind_p   + k*owidth*oheight + i*owidth + j;
+
+          /* compute local max: */
+          int64_t maxindex = -1;
+          scalar_t maxval = -std::numeric_limits<scalar_t>::max();
+          int64_t tcntr = 0;
+          int64_t x,y;
+          for(y = hstart; y < hend; y += dilationH)
+          {
+            for(x = wstart; x < wend; x += dilationW)
+            {
+              tcntr = y*iwidth + x;
+              scalar_t val = *(ip + tcntr);
+              if ((val > maxval) || std::isnan(val))
+              {
+                maxval = val;
+                maxindex = tcntr;
+              }
+            }
+          }
+
+          /* set output to local max */
+          *op = maxval;
+
+          /* store location of max */
+          *indp = maxindex;
+        }
+      }
+    }
+  });
+}
+
+template <typename scalar_t>
+static void max_pool2d_with_indices_out_frame(
+          scalar_t *input_data,
+          scalar_t *output_data,
+          int64_t *indices_data,
+          int64_t nbatch,
+          int64_t nInputPlane,
+          int64_t inputWidth,
+          int64_t inputHeight,
+          int64_t outputWidth,
+          int64_t outputHeight,
+          int kW,
+          int kH,
+          int dW,
+          int dH,
+          int padW,
+          int padH,
+          int dilationW,
+          int dilationH)
+{
+  at::parallel_for(0, nbatch, 0, [&](int64_t start, int64_t end) {
+    for (auto p = start; p < end; p++) {
+      max_pool2d_with_indices_single_out_frame(
+        input_data+p*nInputPlane*inputWidth*inputHeight,
+        output_data+p*nInputPlane*outputWidth*outputHeight,
+        indices_data+p*nInputPlane*outputWidth*outputHeight,
+        nInputPlane,
+        inputWidth, inputHeight,
+        outputWidth, outputHeight,
+        kW, kH, dW, dH,
+        padW, padH,
+        dilationW, dilationH);
+    }
+  });
+}
+
+void max_pool2d_with_indices_out_cpu_template(
+          Tensor& output,
+          Tensor& indices,
+          const Tensor& input_,
+          IntArrayRef kernel_size,
+          IntArrayRef stride,
+          IntArrayRef padding,
+          IntArrayRef dilation,
+          bool ceil_mode)
+{
+  // XXX JIT: Pooling.cpp allows stride.empty().
```

Comment:
This sounds like another misannotated function.

Reply:
The annotation `Optional[BroadcastingList2[int]]` looks correct, but there's some explicit handling of `stride` here:

https://github.com/pytorch/pytorch/blob/70ecddfd76e19e856f2a5aa61c2123eee611af7b/torch/nn/functional.py#L482

I wonder if that could just be:

```
if stride is None:
    stride = kernel_size
```

The drawback is that one could crash `cpp` code by modifying Python code, but that is also the case for the annotation.

Reply:
The other special case is here:

https://github.com/pytorch/pytorch/blob/70ecddfd76e19e856f2a5aa61c2123eee611af7b/test/cpp/api/integration.cpp#L303

It ends up calling `max_pool2d_with_indices` with all default parameters.

I didn't look very hard, but I couldn't find the overload of `torch::max_pool2d(conv1->forward(x), {2, 2})` that works with just two parameters in the `cpp` files.

Reply:
> The drawback is that one could crash cpp code by modifying Python code, but that is also the case for the annotation.

I think, for now, we should consider these annotations as part of the TCB. I do wonder a little if they can't be checked for consistency with the main annotations, but... well... someone would have to figure that out :>

Reply:
> I wonder if that could just be:

Then we'd have to explain why it was permissible for all the other sites to also default things to empty list. Maybe they're all wrong.

@ailzhang, you helped us resolve the annotation last time in #20306. Do you know what's going on with these `Optional[BroadcastingList2]` things?

Reply:
Hi @skrah @ezyang , yea this annotation looks fine, but I guess JIT should have instantiate it to int[2] before passing into aten. I'm looking into it, to double check the repro process:
1. checking out this PR
2. Run IntegrationTest.MNIST should trigger stride.empty() here right?
Thanks!

Reply:
Hi @ailzhang, I think the `stride.empty()` case is also triggered by the regular Python tests if that is more convenient.

`IntegrationTest.MNIST` triggers all of `stride.empty() && padding.size()==1 && dilation.size()==1`.

Reply:
And (for the sake of completeness) this PR already contains a workaround, so `assert()` will need to be added to reproduce it.

Reply:
@ezyang I have a commit to fix this(and other similar ones). Once this PR is merged into master, I will rebase and send it out. :D